infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/ParallelDockerImagesPreparer.java

```diff
@@ -0,0 +1,323 @@
+   * @param machineImageName name of the image that should be assigned on pull
+   * @param progressMonitor consumer of output
+   * @throws SourceNotFoundException if image for pulling not found in registry
+   * @throws InternalInfrastructureException if any other error occurs
+   */
+  protected void pullImage(
+      DockerContainerConfig container, String machineImageName, ProgressMonitor progressMonitor)
+      throws InternalInfrastructureException, SourceNotFoundException {
+    final DockerImageIdentifier dockerImageIdentifier;
+    try {
+      dockerImageIdentifier = DockerImageIdentifierParser.parse(container.getImage());
+    } catch (DockerFileException e) {
+      throw new InternalInfrastructureException(
+          "Try to build a docker machine source with an invalid location/content. It is not in the expected format",
+          e);
+    }
+    if (dockerImageIdentifier.getRepository() == null) {
+      throw new InternalInfrastructureException(
+          format(
+              "Machine creation failed. Machine source is invalid. No repository is defined. Found '%s'.",
+              dockerImageIdentifier.getRepository()));
+    }
+    try {
+      boolean isImageExistLocally =
+          isDockerImageExistLocally(dockerImageIdentifier.getRepository());
+      if (doForcePullImage || !isImageExistLocally) {
+        PullParams pullParams =
+            PullParams.create(dockerImageIdentifier.getRepository())
+                .withTag(MoreObjects.firstNonNull(dockerImageIdentifier.getTag(), LATEST_TAG))
+                .withRegistry(dockerImageIdentifier.getRegistry())
+                .withAuthConfigs(dockerCredentials.getCredentials());
+        dockerConnector.pull(pullParams, progressMonitor);
+      }
+
+      String fullNameOfPulledImage = container.getImage();
+      try {
+        // tag image with generated name to allow sysadmin recognize it
+        dockerConnector.tag(TagParams.create(fullNameOfPulledImage, machineImageName));
+      } catch (ImageNotFoundException nfEx) {
+        throw new SourceNotFoundException(nfEx.getLocalizedMessage(), nfEx);
+      }
+    } catch (IOException e) {
+      throw new InternalInfrastructureException(
+          "Can't create machine from image. Cause: " + e.getLocalizedMessage(), e);
+    }
+  }
+
+  @VisibleForTesting
+  boolean isDockerImageExistLocally(String imageName) {
+    try {
+      return !dockerConnector
+          .listImages(
+              ListImagesParams.create()
+                  .withFilters(new Filters().withFilter("reference", imageName)))
+          .isEmpty();
+    } catch (IOException e) {
+      LOG.warn("Failed to check image {} availability. Cause: {}", imageName, e.getMessage(), e);
+      return false; // consider that image doesn't exist locally
+    }
+  }
+
+  @PreDestroy
+  private void terminate() throws InterruptedException {
+    if (!executor.isShutdown()) {
```

Comment:
@mshaposhnik @skabashnyuk maybe it would be better to make threads daemons?

Reply:
yes

Reply:
so we don't need to shutdown them explicitly in this case ?

Reply:
yep